### ISSUE TITLE
Multiple GPU training on LSTMs

### ISSUE THREAD

**Issue by abhimohta** [Microsoft/CNTK#2446]
By virtue of this - https://github.com/Microsoft/CNTK/issues/2214 I am not being able to use next_minibatch. I want to try and use distributed training on my seq2seq model but I can't use both the ways mentioned on your page(https://docs.microsoft.com/en-us/cognitive-toolkit/Multiple-GPUs-and-machines) training session and next_minibatch because of the above bug.

I am creating my own minibatches like done in https://github.com/Microsoft/CNTK/blob/master/Tutorials/CNTK_106A_LSTM_Timeseries_with_Simulated_Data.ipynb. Is there a way I can implement multiple GPU training with this?

**Comment by KeDengMS** [comment]
Yes, please refer to this [code] (https://github.com/Microsoft/CNTK/blob/nikosk/bidaf/Examples/Text/BidirectionalAttentionFlow/train_pm.py#L200-L214) on how to feed data from python for distributed training. Basically, each worker runs its own process, and it's important to feed data according to its rank. Don't feed the same data to workers during distributed training, as it just effectively bumps up the learning rate.

**Comment by abhimohta** [comment]
I am saving models after every n(=100) epochs. I see 4 models being generated after 100 epochs (num_workers=4) and the performance is pathetic in comparison to the same 100 epoch run without distributed training. Is there something I need to take care while saving models?

**Comment by KeDengMS** [comment]
You have two ways to save model, [Trainer.save_checkpoint](https://cntk.ai/pythondocs/cntk.train.trainer.html?highlight=save_checkpoint#cntk.train.trainer.Trainer.save_checkpoint) or [Function.save](https://cntk.ai/pythondocs/cntk.ops.functions.html?highlight=save#cntk.ops.functions.Function.save). Trainer knows about distributed training, so the actual saving would happen only in rank 0. Function does not, so you need to save only in rank 0 by:

    if 0 == C.Communicator.rank():
        prediction.save('model - %s - %d.cmf' % (guid, cur_epoch))
